p4-farm-nft/src/vault.rs

```diff
@@ -0,0 +1,299 @@
+//! Vault is information per user about their balances in the exchange.
+use crate::*;
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize)]
+#[cfg_attr(feature = "test", derive(Default, Debug, Clone))]
+pub struct Vault {
+    /// Contract.reward_acc value when the last ping was called and rewards calculated
+    pub reward_acc: Balance,
+    /// Staking tokens locked in this vault
+    /// index - contract id
+    /// value - token ids - []
+    pub staked: Vec<TokenIds>,
+    pub min_stake: Balance,
+    /// Amount of accumulated, not withdrawn farmed units. When withdrawing the
+    /// farmed units are translated to all `Contract.farm_tokens` based on
+    /// `Contract.farm_token_rates`
+    pub farmed: Balance,
+    /// farmed tokens which failed to withdraw to the user.
+    pub farmed_recovered: Vec<Balance>,
+    /// NFTs deposited to get an extra boost. Only one NFT can be deposited to a
+    /// single acocunt.
+    /// Storing like `nft_contract@token_id`
+    pub boost_nft: ContractNftTokenId,
+    /// Staked Cheddar. Equals to `Contract.cheddar_rate` * total_staked_tokens.
+    /// not depends on which NFT contract staked more or less tokens, rate used as a const
+    pub total_cheddar_staked: Balance
+}
+
+impl Vault {
+    pub fn new(staked_len: usize, farmed_len: usize, reward_acc: Balance) -> Self {
+        Self {
+            reward_acc,
+            staked: vec![TokenIds::new(); staked_len],
+            min_stake: 0,
+            farmed: 0,
+            farmed_recovered: vec![0; farmed_len],
+            boost_nft: TokenId::new(),
+            total_cheddar_staked: 0
+        }
+    }
+
+    /**
+    Update rewards for locked tokens in past epochs
+    Arguments:
+     - `reward_acc`: Contract.reward_acc value
+     - `round`: current round
+     */
+    pub fn ping(&mut self, reward_acc: Balance, round: u64) {
+        // note: the last round is at self.farming_end
+        // if farming didn't start, ignore the rewards update
+        if round == 0 {
+            return;
+        }
+        // no new rewards
+        if self.reward_acc >= reward_acc {
+            return; // self.farmed;
+        }
+        self.farmed += self.min_stake * (reward_acc - self.reward_acc) / ACC_OVERFLOW;
+        self.reward_acc = reward_acc;
+    }
+
+    /// If all vault's units is empty returns true
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        all_zeros(&self.staked) && self.farmed == 0 && self.boost_nft.is_empty()
+    }
+    /// Get amount of all user NFT tokens staked from all `nft_contract_ids`
+    pub fn get_number_of_staked_tokens(&self) -> usize {
+        let staked_tokens = &self.staked;
+        let mut result:usize = 0;
+        for contract in staked_tokens.iter() {
+            result += contract.len();
+        }
+        result
```

Comment:
```suggestion
        self.staked.iter().sum()
```

Reply:
Actually, this is `Vec<Vec<_>>` but I optimize this through `iter().map()`
https://github.com/YellingOilbird/cheddar/blob/d2f1929e2f8ac966e7546e9732fa38aa17a49037/p4-farm-nft/src/vault.rs#L68